### ISSUE TITLE
NFS browsing (again probably)

### ISSUE THREAD

**Issue by DaVukovic** [sahlberg/libnfs#294]
I know that it might pretty well be related to #232 but still I decided to open a new ticket for that. 

Recently I noticed that I`m no longer able to browse my NFS shares provided by an UnRaid server while using Kodi. I was used to hit that: 

![grafik](https://user-images.githubusercontent.com/7235787/62481252-37d0c000-b7b2-11e9-9a03-1480f519b944.png)

and then I was able to see the IP of my UnRaid server (192.168.1.175). That doesn't happen anymore and all I get is the IP from my Laptop I run Kodi on (192.168.1.182). This laptop is also providing a share. So I was wondering, why do I see my laptop's NFS share, but not the one from my unraid server?!

I did some checks and tried to see which NFS server versions are available on each: 

`rpcinfo -p 192.168.175 | grep -i "nfs"`:
```
davu@davu-Laptop-E-series:~$ rpcinfo -p 192.168.1.175 | grep -i "nfs"
    100003    3   tcp   2049  nfs
    100003    3   udp   2049  nfs
```

`rpcinfo -p 192.168.182 | grep -i "nfs"`:
```
davu@davu-Laptop-E-series:~$ rpcinfo -p 192.168.1.182 | grep -i "nfs"
    100003    3   tcp   2049  nfs
    100003    4   tcp   2049  nfs
    100003    3   udp   2049  nfs
```

So I found the difference, that my UnRaid server is currently not NFSv4 capable in its current state. That never was a problem as I know for sure, that I was able to see both shares (the laptop's and the UnRaid share) while browsing for them. 

As I also read the other ticket and tried different things already, I also tried: 

```
davu@davu-Laptop-E-series:~$ rpcinfo -b 100005 2
192.168.1.182.229.37	davu-Laptop-E-series
192.168.1.182.229.37	davu-Laptop-E-series.fritz.box
192.168.1.182.229.37	davu-Laptop-E-series.fritz.box
192.168.1.182.229.37	davu-Laptop-E-series.fritz.box
davu@davu-Laptop-E-series:~$ 
```

where only my latop is being shown and not a single sign of my UnRaid server. 

So I'm a bit confused what changed lately that I'm not able to browse my NFS shares anymore in the way I was used to. 

I know it's not a problem to let the OS mount the shares and point to that mountpoint from inside Kodi. But explaining that to the public joe who uses Kodi for the very first time and somehow got it managed to get a NFS-share up and running might be a bit too hard for them ;)

Btw...using the "add network location", selecting "NFS" and enter the IP andthe path to that share (found out by `showmount -e <IP>` for example) works like a charm. So it's not the case that Kodi can't use NFSv3 anymore. It just dosn't show up while browsing

Sorry for the long text. If you need any additional info, please let me know

Thanks in advance.

**Comment by sahlberg** [comment]
libnfs uses the same mechanism as "

rpcinfo -b 100005 2

" to discover nfs servers.

This utilizes broadcasts on the UDP protocol to the portmapper and the
mount daemon.

Check with rpcinfo that the both portmapper and the mount daemon is running.
(portmapper is running or else rpcinfo wouldnt work.)

For the mount daemon you should see something like :
...
    100005    2    udp       0.0.0.0.78.80          mountd     superuser
    100005    2    tcp       0.0.0.0.78.80          mountd     superuser
    100005    2    udp6      ::.78.80               mountd     superuser
    100005    2    tcp6      ::.78.80               mountd     superuser
    100005    3    udp       0.0.0.0.78.80          mountd     superuser
    100005    3    tcp       0.0.0.0.78.80          mountd     superuser
    100005    3    udp6      ::.78.80               mountd     superuser
    100005    3    tcp6      ::.78.80               mountd     superuser
...

If it is running, then check that you don't have any firewall rules on
client or server that would drop UDP packets.

Run a wireshark trace when you are using "rpcinfo -b 100005 2" and see if
the server responds or not.
If it does not respond to any of the CALLIT commands that are sent out then
it is a server issue.

**Comment by DaVukovic** [comment]
Using wireshark is a bit over my paygrade, tbh. I tried using it while sending `rpcinfo -p 192.168.1.175` which is my Unraid server then, I got the follwing: 

![image](https://user-images.githubusercontent.com/7235787/62640487-fe2ebf00-b941-11e9-9805-6a4ac39ed0bd.png)

That's the output of the command above: 

```
davu@davu-Laptop-E-series:~$ rpcinfo -p 192.168.1.175
   program vers proto   port  service
    100000    4   tcp    111  portmapper
    100000    3   tcp    111  portmapper
    100000    2   tcp    111  portmapper
    100000    4   udp    111  portmapper
    100000    3   udp    111  portmapper
    100000    2   udp    111  portmapper
    100024    1   udp  56618  status
    100024    1   tcp  58761  status
    100003    3   tcp   2049  nfs
    100003    3   udp   2049  nfs
    100021    1   udp  56916  nlockmgr
    100021    3   udp  56916  nlockmgr
    100021    4   udp  56916  nlockmgr
    100021    1   tcp  33243  nlockmgr
    100021    3   tcp  33243  nlockmgr
    100021    4   tcp  33243  nlockmgr
    100005    1   udp  38579  mountd
    100005    1   tcp  60311  mountd
    100005    2   udp  51153  mountd
    100005    2   tcp  39953  mountd
    100005    3   udp  35104  mountd
    100005    3   tcp  51367  mountd
davu@davu-Laptop-E-series:~$ 
```

and to compare, from localhost: 
```
davu@davu-Laptop-E-series:~$ rpcinfo -p localhost
   program vers proto   port  service
    100000    4   tcp    111  portmapper
    100000    3   tcp    111  portmapper
    100000    2   tcp    111  portmapper
    100000    4   udp    111  portmapper
    100000    3   udp    111  portmapper
    100000    2   udp    111  portmapper
    100005    1   udp  43761  mountd
    100005    1   tcp  49615  mountd
    100005    2   udp  47349  mountd
    100005    2   tcp  47919  mountd
    100005    3   udp  39845  mountd
    100005    3   tcp  33179  mountd
    100003    3   tcp   2049  nfs
    100003    4   tcp   2049  nfs
    100227    3   tcp   2049
    100003    3   udp   2049  nfs
    100227    3   udp   2049
    100021    1   udp  47200  nlockmgr
    100021    3   udp  47200  nlockmgr
    100021    4   udp  47200  nlockmgr
    100021    1   tcp  36877  nlockmgr
    100021    3   tcp  36877  nlockmgr
    100021    4   tcp  36877  nlockmgr
davu@davu-Laptop-E-series:~$ 
```

I don't have any kind of firewalls running. Neither on the Ubuntu machine nor on UnRaid. 

I also tried to let wireshark capture what happens by sending `rpcinfo -b ......` and I don't see 192.168.1.175 responding. Only 192.168.1.182 (my laptop) is responding.

So that's a UnRaid issue then?

**Comment by sahlberg** [comment]
On Thu, Aug 8, 2019 at 2:37 AM DaVukovic <notifications@github.com> wrote:

> Using wireshark is a bit over my paygrade, tbh. I tried using it while
> sending rpcinfo -p 192.168.1.175 which is my Unraid server then, I got
> the follwing:
>
> [image: image]
> <https://user-images.githubusercontent.com/7235787/62640487-fe2ebf00-b941-11e9-9805-6a4ac39ed0bd.png>
>
> That's the output of the command above:
>
> davu@davu-Laptop-E-series:~$ rpcinfo -p 192.168.1.175
>    program vers proto   port  service
>     100000    4   tcp    111  portmapper
>     100000    3   tcp    111  portmapper
>     100000    2   tcp    111  portmapper
>     100000    4   udp    111  portmapper
>     100000    3   udp    111  portmapper
>     100000    2   udp    111  portmapper
>     100024    1   udp  56618  status
>     100024    1   tcp  58761  status
>     100003    3   tcp   2049  nfs
>     100003    3   udp   2049  nfs
>     100021    1   udp  56916  nlockmgr
>     100021    3   udp  56916  nlockmgr
>     100021    4   udp  56916  nlockmgr
>     100021    1   tcp  33243  nlockmgr
>     100021    3   tcp  33243  nlockmgr
>     100021    4   tcp  33243  nlockmgr
>     100005    1   udp  38579  mountd
>     100005    1   tcp  60311  mountd
>     100005    2   udp  51153  mountd
>     100005    2   tcp  39953  mountd
>     100005    3   udp  35104  mountd
>     100005    3   tcp  51367  mountd
> davu@davu-Laptop-E-series:~$
>
> and to compare, from localhost:
>
> davu@davu-Laptop-E-series:~$ rpcinfo -p localhost
>    program vers proto   port  service
>     100000    4   tcp    111  portmapper
>     100000    3   tcp    111  portmapper
>     100000    2   tcp    111  portmapper
>     100000    4   udp    111  portmapper
>     100000    3   udp    111  portmapper
>     100000    2   udp    111  portmapper
>     100005    1   udp  43761  mountd
>     100005    1   tcp  49615  mountd
>     100005    2   udp  47349  mountd
>     100005    2   tcp  47919  mountd
>     100005    3   udp  39845  mountd
>     100005    3   tcp  33179  mountd
>     100003    3   tcp   2049  nfs
>     100003    4   tcp   2049  nfs
>     100227    3   tcp   2049
>     100003    3   udp   2049  nfs
>     100227    3   udp   2049
>     100021    1   udp  47200  nlockmgr
>     100021    3   udp  47200  nlockmgr
>     100021    4   udp  47200  nlockmgr
>     100021    1   tcp  36877  nlockmgr
>     100021    3   tcp  36877  nlockmgr
>     100021    4   tcp  36877  nlockmgr
> davu@davu-Laptop-E-series:~$
>
> I don't have any kind of firewalls running. Neither on the Ubuntu machine
> nor on UnRaid.
>
> I also tried to let wireshark capture what happens by sending rpcinfo -b
> ...... and I don't see 192.168.1.175 responding. Only 192.168.1.182 (my
> laptop) is responding.
>
> So that's a UnRaid issue then?
>

Yepp. if it doesn't respond then it is an unraid issue.

**Comment by DaVukovic** [comment]
Ok, thanks. I'll report to them and keep you noted if you are interested

**Comment by sahlberg** [comment]
Sure, if you hear from them and if they have a fix or if there are only
specific version that are affected, or if they have a solution,
please let me know and I can add it to my README file in case other users
encounter it too.

**Comment by MilhouseVH** [comment]
I see a similar issue with FreeNAS 8.3 although this is a very old version and I do need to upgrade (it being a NAS - which "just works" - it's not really something I want to tinker with until absolutely necessary!!)

My setup: I have a NFS share on Ubuntu 17.10 (which I think is NFSv3/NFSv4), and NFS shares on FreeNAS 8.3 (based on FreeBSD 8.3) which I believe are NFSv3 only. These are the responses to `rpcinfo -b 100005 #` where `#` (`versnum`) is 1, 2, 3 when run from Ubuntu 17.10 (there is no firewall etc.):

Ubuntu: `nm-linux.local` (192.168.0.9)
FreeNAS: `freenas.local` (192.168.0.3)

```
neil@nm-linux:~$ rpcinfo -b 100005 1
192.168.0.9.199.133     nm-linux.local
192.168.0.9.199.133     nm-linux.local
192.168.0.3.3.240       freenas.local
192.168.0.3.3.240       freenas.local
192.168.0.9.199.133     nm-linux.local
192.168.0.9.199.133     nm-linux.local
192.168.0.3.3.240       freenas.local
192.168.0.3.3.240       freenas.local
```
```
neil@nm-linux:~$ rpcinfo -b 100005 2
192.168.0.9.214.255     nm-linux.local
192.168.0.9.214.255     nm-linux.local
192.168.0.9.214.255     nm-linux.local
192.168.0.9.214.255     nm-linux.local
```
```
neil@nm-linux:~$ rpcinfo -b 100005 3
192.168.0.9.189.197     nm-linux.local
192.168.0.9.189.197     nm-linux.local
192.168.0.3.3.240       freenas.local
192.168.0.3.3.240       freenas.local
192.168.0.9.189.197     nm-linux.local
192.168.0.9.189.197     nm-linux.local
192.168.0.3.3.240       freenas.local
192.168.0.3.3.240       freenas.local
```

(no response at all for `versnum`=`4`)

So we can see that Ubuntu responds to `versnum` 1, 2 and 3, but FreeNAS responds only to `versnum` 1 and 3, which means it isn't discovered by libnfs (in LibreELEC/Kodi, I'm only able to discover the Ubuntu share).

I don't think this is a new issue, I believe FreeNAS has always had this "quirk" - maybe it's fixed in more modern versions of FreeBSD (one day, I'll find out...) and maybe it also affects UnRaid. @sahlberg any ideas why FreeNAS (like UnRaid) wouldn't respond to `versnum=2`, or how (even if) `versnum` relates to NFS version?

**Comment by MilhouseVH** [comment]
Just for reference:

UnRaid: https://forums.unraid.net/bug-reports/stable-releases/670-nfs-r551/

FreeNAS: https://www.ixsystems.com/community/threads/freenas-nfs-not-discovered-by-machines-on-the-network.60079/

And a summary of `rpcinfo` as seen from my Ubuntu PC:
```
neil@nm-linux:~$ rpcinfo -s nm-linux
   program version(s) netid(s)                         service     owner
    100000  2,3,4     local,udp,tcp,udp6,tcp6          portmapper  superuser
    100005  3,2,1     tcp6,udp6,tcp,udp                mountd      superuser
    100003  4,3       udp6,tcp6,udp,tcp                nfs         superuser
    100227  3         udp6,tcp6,udp,tcp                -           superuser
    100021  4,3,1     tcp6,udp6,tcp,udp                nlockmgr    superuser
    100024  1         tcp6,udp6,tcp,udp                status      123
```
```
neil@nm-linux:~$ rpcinfo -s freenas
   program version(s) netid(s)                         service     owner
    100000  2,3,4     local,udp6,tcp6,udp,tcp          portmapper  superuser
    100005  3,1       tcp,udp,tcp6,udp6                mountd      superuser
    100003  3,2       tcp6,tcp,udp6,udp                nfs         superuser
    100024  1         tcp,udp,tcp6,udp6                status      superuser
    100021  4,3,1,0   tcp,udp,tcp6,udp6                nlockmgr    superuser
```

Totally accept this is a server issue (UnRaid, FreeNAS etc.) and maybe one day the developers of those distributions will get to the bottom of why there is no response to `rpcinfo -b 100005 2`.

**Comment by DaVukovic** [comment]
>and maybe one day the developers of those distributions will get to the bottom of why there is no response to rpcinfo -b 100005 2.

Well, at least for UnRaid 6.5.x it worked fine. That's the version I upgraded from to the current 6.7.x. So something seems to have changed for whatever reason. I hope they will react on that as they labeled it with `minor` priority.

**Comment by sahlberg** [comment]
On Fri, Aug 9, 2019 at 4:31 AM MilhouseVH <notifications@github.com> wrote:

> Just for reference:
>
> UnRaid:
> https://forums.unraid.net/bug-reports/stable-releases/670-nfs-r551/
>
> FreeNAS:
> https://www.ixsystems.com/community/threads/freenas-nfs-not-discovered-by-machines-on-the-network.60079/
>
> And a summary of rpcinfo as seen from my Ubuntu PC:
>
> neil@nm-linux:~$ rpcinfo -s nm-linux
>    program version(s) netid(s)                         service     owner
>     100000  2,3,4     local,udp,tcp,udp6,tcp6          portmapper  superuser
>     100005  3,2,1     tcp6,udp6,tcp,udp                mountd      superuser
>     100003  4,3       udp6,tcp6,udp,tcp                nfs         superuser
>     100227  3         udp6,tcp6,udp,tcp                -           superuser
>     100021  4,3,1     tcp6,udp6,tcp,udp                nlockmgr    superuser
>     100024  1         tcp6,udp6,tcp,udp                status      123
>
> neil@nm-linux:~$ rpcinfo -s freenas
>    program version(s) netid(s)                         service     owner
>     100000  2,3,4     local,udp6,tcp6,udp,tcp          portmapper  superuser
>     100005  3,1       tcp,udp,tcp6,udp6                mountd      superuser
>     100003  3,2       tcp6,tcp,udp6,udp                nfs         superuser
>     100024  1         tcp,udp,tcp6,udp6                status      superuser
>     100021  4,3,1,0   tcp,udp,tcp6,udp6                nlockmgr    superuser
>
> Totally accept this is a server issue (UnRaid, FreeNAS etc.) and maybe one
> day the developers of those distributions will get to the bottom of why
> there is no response to rpcinfo -b 100005 2.
>

Actually, Unraid no longer support mount version 2.
Can you try  "rpcinfo -b 100005 3"
and see if that works, in that case we can possibly fix this in the short
term.

Longer term, the outlook for discovery is bleak. Once NFSv4-only servers
become common, this will not work since they will not have a portmapper to
broadcast to.
Additionally, my newly installed Fedora30 server does not respond to callit
at all, so I imagine they are removing these things as time goes by.
rpcinfo -b   and libnfs are very likely the only two pprograms in the world
that actually use the CALLIT procedure that this type of discovery builds
on.
So pure speculation, it might have been removed serverside to fix an issue
or to remove "dead" code to reduce maintenance cost or something else.

Longer term, I think a day will come where this type of discovery will just
not work anymore for NFS. :-(
There are alternative ways to try to discover NFS servers. We could just
iterate over every host in the subnet and send a direct call to port 2049
for the RPC NULL procedure and see if we get a reply.  But then we are
basically starting to behave like a port-scanner and flood the networks
everytime we try to find nfs server.
I am not sure that is a good idea.

**Comment by DaVukovic** [comment]
>Can you try  "rpcinfo -b 100005 3"
and see if that works, in that case we can possibly fix this in the short
term.

Will try later this evening and will report back

**Comment by DaVukovic** [comment]
Sorry for the delay

`rpcinfo -b 100005 3` doesn't give me any output while my UnRaid is the only device sharing something.
